packages/marko/src/runtime/components/init-components-browser.js

```diff
@@ -252,41 +251,38 @@ function initClientRendered(componentDefs, doc) {
  * This method initializes all components that were rendered on the server by iterating over all
  * of the component IDs.
  */
-function initServerRendered(renderedComponents, doc) {
+function initServerRendered(renderedComponents, doc, runtimeId) {
   var type = typeof renderedComponents;
-  var globalKey = "$";
-  var runtimeId;
+  var globalKey;
 
   if (type !== "object") {
     if (type === "string") {
       runtimeId = renderedComponents;
-      globalKey += runtimeId + "_C";
+      globalKey = runtimeId;
     } else {
-      globalKey += (runtimeId = DEFAULT_RUNTIME_ID) + "C";
+      globalKey = runtimeId = DEFAULT_RUNTIME_ID;
     }
 
-    renderedComponents = win[globalKey];
-
-    var fakeArray = (win[globalKey] = {
-      r: runtimeId,
-      concat: initServerRendered
+    renderedComponents = JSON[globalKey];
+    Object.defineProperty(JSON, globalKey, {
+      configurable: true,
+      set: function(v) {
+        initServerRendered(JSON.parse(v.slice(10)), doc, runtimeId);
+      }
     });
 
-    if (renderedComponents && renderedComponents.forEach) {
-      renderedComponents.forEach(function(renderedComponent) {
-        fakeArray.concat(renderedComponent);
+    if (renderedComponents) {
+      JSON.parse("[" + renderedComponents.slice(10) + "]").forEach(function(
```

Comment:
I like this — it looks like in the future it could even throttle itself based on how much distress the user’s CPU is in

Reply:
Yeah that was one of the things I liked as well, we can lazily parse the JSON.